### PR TITLE
Implement label smoothing for CrossEntropy Loss

### PR DESCRIPTION
**Pull request by dbogunowicz**
CrossEntropy losses (Categorical) now can take label smoothing are as parameter.
In theory, ready to be reviewed, but for clarity, I will open this PR once the PyTorch 1.10.2 upgrade gets merged to main (dependency).

### CODE REVIEW

File: tests/sparseml/pytorch/utils/test_loss.py [hunk 1]
```diff
@@ -64,6 +64,28 @@ def default_kd_parent():
 DEFAULT_KD_SETTINGS = KDSettings(teacher=default_kd_parent())
 
 
+@pytest.mark.parametrize(
```
Comment (dbogunowicz) [inline tests/sparseml/pytorch/utils/test_loss.py:67]: I am not very proud of those tests, will refactor it in the convenient moment.

File: src/sparseml/pytorch/utils/loss.py [hunk 2]
```diff
@@ -213,10 +213,36 @@ class CrossEntropyLossWrapper(LossWrapper):
 
     def __init__(
         self,
+        label_smoothing: float = 0.0,
```
Comment (bfineran) [inline src/sparseml/pytorch/utils/loss.py:216]: will this break on `torch<1.10`?

**Review by bfineran — CHANGES_REQUESTED** [review]
looks nice and clean, can you change the base of this branch to the 10.2 update branch to make it easier to do a full review? additionally, note that `integrations/pytorch` is currently being deprecated in favor of `src/pytorch/image_classification`

File: src/sparseml/pytorch/utils/loss.py (hunk 2 above)
Comment (dbogunowicz) [reply src/sparseml/pytorch/utils/loss.py:216]: To be exact, line `230` (`self._loss_fn`) will throw an error, because on lower pytorch versions the loss function will not recognise the `label_smoothing` argument.

Comment (bfineran) [reply src/sparseml/pytorch/utils/loss.py:216]: let's definitely make sure this is compatible with lower torch versions (ie check torch version and only include label smoothing if >=1.10 and raise an error if torch < 1.10 and label_smoothing != 0)

**Comment by bfineran** [comment]
@dbogunowicz looks like there's still some conflicts/unintended changes. can you take a look?

File: src/sparseml/pytorch/image_classification/train.py [hunk 3]
```diff
@@ -512,10 +524,18 @@ def train(
     :param loggers: List of loggers to use during training process
     """
     # loss setup
-    val_loss = helpers.get_loss_wrapper(arch_key=train_args.arch_key, training=True)
+    val_loss = helpers.get_loss_wrapper(
+        arch_key=train_args.arch_key,
+        loss_func_args=train_args.loss_func_args,
+        training=False,
```
Comment (dbogunowicz) [inline src/sparseml/pytorch/image_classification/train.py:530]: I additionally changed `training=False`, it seems to be a logical argument here.

File: integrations/pytorch/train.py [hunk 4]
```diff
@@ -513,10 +526,18 @@ def train(
     :param loggers: List of loggers to use during training process
     """
     # loss setup
-    val_loss = utils.get_loss_wrapper(arch_key=train_args.arch_key, training=True)
+    val_loss = utils.get_loss_wrapper(
+        arch_key=train_args.arch_key,
+        loss_func_args=train_args.loss_func_args,
+        training=False,
```
Comment (dbogunowicz) [inline integrations/pytorch/train.py:532]: I additionally changed `training=False`, it seems to be a logical argument here.

File: src/sparseml/pytorch/utils/loss.py (hunk 2 above)
Comment (dbogunowicz) [reply src/sparseml/pytorch/utils/loss.py:217]: Refactored.

**Comment by rahul-tuli** [comment]
LGTM pending merge conflicts

File: src/sparseml/pytorch/image_classification/utils/helpers.py [hunk 5]
```diff
@@ -302,11 +304,15 @@ def get_loss_wrapper(
...
-            CrossEntropyLossWrapper(extras)
+            CrossEntropyLossWrapper(
+                label_smoothing=loss_func_args["label_smoothing"], extras=extras
+            )
             if training and "inception" not in arch_key
             else InceptionCrossEntropyLossWrapper(extras)
         )
-    return LossWrapper(loss_fn=torch_functional.cross_entropy, extras=extras)
+    return CrossEntropyLossWrapper(
+        label_smoothing=loss_func_args["label_smoothing"], extras=extras
```
Comment (rahul-tuli) [inline src/sparseml/pytorch/image_classification/utils/helpers.py:314]: Same as comment on line 308

File: src/sparseml/pytorch/image_classification/train.py [hunk 6]
```diff
@@ -512,10 +524,18 @@ def train(
     :param loggers: List of loggers to use during training process
     """
     # loss setup
-    val_loss = helpers.get_loss_wrapper(arch_key=train_args.arch_key, training=True)
```
Comment (rahul-tuli) [inline src/sparseml/pytorch/image_classification/train.py:515]: This func needs to be updated?
